### ISSUE TITLE
Back button doesn't change the url when state changed

### ISSUE THREAD

**Issue by tux240** [angular-ui/ui-router#3040]
I'm using "stateChangeStart" in order to refresh the informations of the connected user .
In the callback, there is a state.go with the url.name (asked state), the parameters and some options.
It work like a charm.

However, when I use the back button, the previous state is displayed but the url doesn't change.
Options are:
$retry:false
inherit:true
location:false
notify:true
relative:null
reload:false

Do you have some suggestions ?

Thanks

**Comment by christopherthielen** [comment]
There's not enough information given. 

This tracker is for issues only.  Generally, we send people to StackOverflow for support requests http://bit.ly/UIR-SOF ... However, if you provide a minimal reproduction, I'd be happy to take a look http://bit.ly/UIR-Plunk

**Comment by tux240** [comment]
Here my part of code where I'm reloading user's data on each new state requested:

`
$rootScope.$on('$stateChangeStart', function (evt, to, toParams, from, fromParams, options) {
        if (!to.$$isUserLoaded) {
          evt.preventDefault();
          $rootScope.refeshCurrentUser().then(function () {
            angular.extend(to, { $$isUserLoaded: true });
            $state.go(to.name, toParams, options).finally(function () { to.$$isUserLoaded = false; });
          });

          return;
        }

        if (!isAuthorized(to)) {
          evt.preventDefault();
          $state.go(getAuthRedirect(to), toParams);
          return;
        }
});
`

When I try to go back, url isn't changed but the view, yes.

**Comment by christopherthielen** [comment]
@tux240 I need you to reproduce the issue (preferably in a plunker) in order to help.  Put the code into a plunker which reproduces the issue.  

I provided a link http://bit.ly/UIR-Plunk to get you started.

**Comment by tux240** [comment]
Ok thanks @christopherthielen 

Here the link to the Plunkr:
http://plnkr.co/edit/qRmnmlg1euyWrpvze6nD?p=preview

You have to go to home, home.foo and home.foo.detail for example.
Next, you have to try to back to the previous view then, The view is changed but not the url !

**Comment by tux240** [comment]
Do you have some suggestions ?

**Comment by christopherthielen** [comment]
@tux240 there's definitely something a little weird going on here.  The console is immediately showing an infinite digest and state changes are happening really slowly.  I'm not immediately sure what's happening

**Comment by christopherthielen** [comment]
I see that you snuck your `$stateChangeStart` handler into `_extras.js`... I hadn't noticed it there.

This does look like a problem in version 0.3.1, but we've created the `Transition` api in 1.0 to address issues like this.  

Here's a plunker that demonstrates how to use the `$transitions` API in 1.0 to implement the code based on your example plunker.  Note that the back button works as normal.

http://plnkr.co/edit/XHvBoF9sCcT2iZZuUIAz?p=preview

However, your example code suggests that you *always* want to invoke the call to `refeshCurrentUser()` on *every* state change to check if they are 1) authenticated and 2) have the rights to access the state.  Transition hooks are designed to do exactly this.

```
// invoke the hook for every transition 
// (criteria object {} matches every transition)
$transitions.onStart({}, function(transition) {
  // Pause transition until the promise settles
  // Refresh the user from the server
  return $rootScope.refeshCurrentUser().then(function() {
    var to = transition.to();
    if (!isAuthorized(to)) {
      // if they're not authorized, redirect to the proper state
      return $state.target(getAuthRedirect(to), transition.params());
    }
  });
});
```


Closing, because I think I've answered how this is fixed by ui-router 1.0
